### PR TITLE
feat: Annotation to support per app refresh timeouts (soft & hard) - #9499

### LINKED ISSUE #9499: Add refreshPolicy for argocd application object a.k.a. (hard-)refresh-policy
# Summary

Add a refresh policy setting to argocd application objects (like syncPolicy, but specifically for refresh strategies).

# Motivation

This applies to many scenarios: tracking branches, head, overridden helm packages etc. I would like to set whether an argocd application does soft or hard refresh by default. Additionally setting the refresh related TTL from the app would also be useful.

### Example:
Rapid development in "snapshot" mode, override package until release and promotion is done.
--> this means that the oci helm coordinates stay the same, but the oci image changes. 
--> kubernetes deals with this by having an imagePullPolicy that you can set to Always, for constant hash verification and pull on difference.

### PR DESCRIPTION
**Pull request by 4ch3los**
Hey together,

with this pr i added two new annotations to control the refresh timeouts(soft and hard) on a per app basis. This should solve the issue https://github.com/argoproj/argo-cd/issues/9499 and also allow users to specify hard refresh intervals, when using plugins like argocd-vault-plugin, to refresh changed secrets(https://argocd-vault-plugin.readthedocs.io/en/stable/usage/#refreshing-values-from-secrets-managers)

Fixes #9499

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [x] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [ ] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [x] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

### CODE REVIEW

**Review by jannfis — COMMENTED** [review]
Thanks for your PR.

I tend to think that giving users control over the refresh time would not be a good idea, as this will put unexpected strain on the application controller and repository server - especially in the case of a hard refresh.

I'm also not sure yet whether a setting such as this should be an annotation, or whether it is better housed in the spec.

IMHO, we should think this through, and come up with a proper way to allow/disallow users from actually leveraging this setting (e.g. by having a flag in the AppProject).

TL;DR: I think this needs further discussion.

**Comment by 4ch3los** [comment]
> Thanks for your PR.
> 
> I tend to think that giving users control over the refresh time would not be a good idea, as this will put unexpected strain on the application controller and repository server - especially in the case of a hard refresh.
> 
> I'm also not sure yet whether a setting such as this should be an annotation, or whether it is better housed in the spec.
> 
> IMHO, we should think this through, and come up with a proper way to allow/disallow users from actually leveraging this setting (e.g. by having a flag in the AppProject).
> 
> TL;DR: I think this needs further discussion.

Hey! I absolutely understand that having to low timeouts will harm performance, but at the moment, the user only has the option to set it on the global level(https://github.com/argoproj/argo-cd/blob/master/cmd/argocd-application-controller/commands/argocd_application_controller.go#L235) which can even more easily harm performance

Regarding the way of implementation, i just used annotations cause there was already a annotation for forcing a refresh, but i can also adjust it and define it via spec :)

**Comment by jannfis** [comment]
I think there is a subtle difference: It's not the user who has control over the global setting, but the Argo CD administrator or operator. With an annotation on the application, that control is handed over to a regular, low-privileged user.

**Review by agaudreault — CHANGES_REQUESTED** [review]
I also think this needs further discussion. I agree with @jannfis on the different arguments above. Argo CD is a GitOps tool and react to git changes.

I think if there is a need to monitor an external non-git systems, like an artifact registry to detect non-gitops changes, this can be done by another controller. Then that controller can request a refresh of the application with the current annotation/api when it detects changes that would require a refresh. A naive implementation would be to do it periodically

**Comment by 4ch3los** [comment]
> I think there is a subtle difference: It's not the user who has control over the global setting, but the Argo CD administrator or operator. With an annotation on the application, that control is handed over to a regular, low-privileged user.

Ahh yeah, our perspective as serviceprovider, where we use argocd to baseline and are admin & user, there was not a big difference. But a global setting to limit the options or disable it overall should be usefull.



> I also think this needs further discussion. I agree with @jannfis on the different arguments above. Argo CD is a GitOps tool and react to git changes.
> 
> I think if there is a need to monitor an external non-git systems, like an artifact registry to detect non-gitops changes, this can be done by another controller. Then that controller can request a refresh of the application with the current annotation/api when it detects changes that would require a refresh. A naive implementation would be to do it periodically

For issues like the linked, this could definitely be a option. Regarding the second usecase with config management plugins(like vault secret plugin), where the same source can result in a different result based on external conditions, this would need a whole restructuring of the plugin design from my perspective.

File: controller/appcontroller.go [hunk 1]
```diff
@@ -1771,6 +1771,21 @@ func (ctrl *ApplicationController) needRefreshAppStatus(app *appv1.Application,
 	compareWith := CompareWithLatest
 	refreshType := appv1.RefreshTypeNormal
 
+	if val, ok := app.Annotations[appv1.AnnotationAppRefreshTimeout]; ok {
```
Comment (andrii-korotkov-verkada) [inline controller/appcontroller.go:1774]: This is better to compute on a higher-level function and keep this function just checking things for the specified params. Overriding passed in params is confusing.
